pkg/transport/listener.go

```diff
@@ -174,6 +178,22 @@ func (info TLSInfo) baseConfig() (*tls.Config, error) {
 		MinVersion:   tls.VersionTLS12,
 		ServerName:   info.ServerName,
 	}
+
+	if info.AllowedCN != "" {
+		cfg.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			for _, chains := range verifiedChains {
+				for _, chain := range chains {
+					if info.AllowedCN == chain.Subject.CommonName {
+						return nil
+					}
+
+					return fmt.Errorf("CommonName authentication failed (allowed: %s, client: %s)", info.AllowedCN, chains[0].Subject.CommonName)
```

Comment:
Just remove this `return fmt.Errorf` entirely because it's possible that one of the chain has different `Subject.CommonName`?

If we still need to keep this error, it should be only when `chain.Subject.CommonName != ""` and also `s/chains[0].Subject.CommonName/chain.Subject.CommonName/`.

Reply:
OK I'll change the behaviour for checking a leaf of every verified chain. Probably checking non-leaf certs should be avoided because they wouldn't be user supplied. If the parameter passed with `--peer-cert-allowed-cn` and CNs in the non leaf certs match accidentally, a node can be authenticated in a wrong manner.

I'm still not sure how the case of multiple verified chains can happen yet... I'm digging it.